### PR TITLE
added support of slicing for IDSStructArray

### PR DESCRIPTION
**Pull request by prasad-sawantdesai**
Added support of slicing for IDSStructArray
```
import imas

entry = imas.DBEntry("imas:mdsplus?user=public;pulse=116000;run=4;database=ITER_MD;version=3", "r")  
walids = entry.get("wall", autoconvert=False) # lazy=True
units = walids.description_2d[0].vessel.unit
print(units[8:])
print(units[0:100])
print(units[4:])
print(units[5])
```

### CODE REVIEW

File: imas/ids_struct_array.py [hunk 1]
```diff
@@ -121,12 +122,25 @@ def _element_structure(self):
...
-            self._load(item)
-        return self.value[list_idx]
+        # allow slices
+        if isinstance(item, slice):
+            if self._lazy:
+                start, stop, step = item.start, item.stop, item.step
+                if stop is None:
+                    stop = sys.maxsize
+
+                for i in range(start or 0, stop, step or 1):
```
Comment (maarten-ic) [inline imas/ids_struct_array.py:132]: Hi Prasad, Python slices are weird and have lots of edge cases with negative indices... the current logic doesn't quite capture that.

For example:
```python
>>> import imas
>>> entry = imas.DBEntry("imas:hdf5?path=./test", "w")
09:44:08 INFO     Parsing data dictionary version 4.0.0 @dd_zip.py:166
>>> cp = entry.factory.core_profiles()
>>> cp.profiles_1d.resize(10)
>>> cp.ids_properties.homogeneous_time = 1
>>> cp.time = [*range(10)]
09:45:05 INFO     Assigning incorrect type 'int64' to <IDSNumericArray (IDS:core_profiles, time, empty FLT_1D)>, attempting automatic conversion. @ids_primitive.py:483
>>> entry.put(cp)
>>> cp.profiles_1d[:-5]
[<IDSStructure (IDS:core_profiles, profiles_1d[0])>, <IDSStructure (IDS:core_profiles, profiles_1d[1])>, <IDSStructure (IDS:core_profiles, profiles_1d[2])>, <IDSStructure (IDS:core_profiles, profiles_1d[3])>, <IDSStructure (IDS:core_profiles, profiles_1d[4])>]
>>> entry.get("core_profiles", lazy=True).profiles_1d[:-5]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/maarten/projects/iter-python/imas-python/imas/ids_struct_array.py", line 137, in __getitem__
    return self.value[item]
           ~~~~~~~~~~^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

Edge cases with negative indices:
- Negative start, e.g. `[-5:5]`, all items from the fifth from the end to the fifth from the beginning
- Negative stop, e.g. `[2:-2]`, all items from the third to the second-to-last
- Negative increment, e.g. `[::-1]` all items in reverse order

What is the the use case for this feature? With a list comprehension users could capture (most) scenarios as well:
```python
p1ds = [cp.profiles_1d[i] for i in range(2, 5)]
# in reverse
for p1d in reversed(cp.profiles_1d):
    ...
```

Comment (prasad-sawantdesai) [reply imas/ids_struct_array.py:132]: Thank you for reviewing @maarten-ic 

Actually, this issue arises when using slicing for wall IDs, as shown below.  There is a workaround—you can convert it into a list—but then you lose the IDSStructArray metadata. It would be convenient for users to work it as a list, what do you think @olivhoenen but of course we need to consider edge cases. 
units = list(walids.description_2d[0].vessel.unit)

```python

import imaspy as imas

entry = imas.DBEntry("imas:mdsplus?user=public;pulse=116000;run=4;database=ITER_MD;version=3", "r")  
walids = entry.get("wall", autoconvert=False)
units = walids.description_2d[0].vessel.unit
# units = list(walids.description_2d[0].vessel.unit) workaround
print(units[8:])

# 09:01:55 INFO     Parsing data dictionary version 4.0.0 @dd_zip.py:166
# 09:01:59 INFO     Parsing data dictionary version 3.40.0 @dd_zip.py:166
# Traceback (most recent call last):
#  File "/home/ITER/sawantp1/git/idstoolsimaspy/idstools/testbug.py", line 6, in <module>
#    print(units[8:])
#         ~~~~~^^^^
#  File "/work/imas/opt/EasyBuild/software/IMASPy/1.2.0-intel-2023b/lib/python3.11/site-
# packages/imaspy/ids_struct_array.py", line 126, in __getitem__
#    list_idx = int(item)
               ^^^^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'slice'
```

Comment (maarten-ic) [reply imas/ids_struct_array.py:132]: Hi Prasad,

> you can convert it into a list—but then you lose the IDSStructArray metadata

I don't understand the drawback: with the slicing you have implemented, `unit[8:]` is also a list, right?

> when using slicing for wall IDs, as shown below

Yes, sure, but what's the actual use case for doing this? 🙂 I don't understand why a user would want to print all units except the first 8 (as you do in your code listing).

Note that I found out that slice objects have a method that give you the start/stop/step values for a sequence of given length: https://docs.python.org/3/reference/datamodel.html#slice.indices

Comment (olivhoenen) [reply imas/ids_struct_array.py:132]: This came up in the context of @prasad-sawantdesai work on IDStools, where different scripts are receiving IMAS URIs that may contain fragments (`#ids:occ/ids_path`) where the `ids_path` has slicing (as described in https://imas-data-dictionary.readthedocs.io/en/latest/_downloads/9e7ffd162c0237b61062528affd5fe2a/IDS-path-syntax.md).

What would be the easiest/recommended way to go from an `ids_path` string that contains slicing indices to an actual list of objects? 

ps: in the case of the list of `unit` from the wall IDS, the machine description contains many different types of `limiter` and `vessel` units, not all of them make sense to overlay when plotting for example an equilibrium psi map

Comment (maarten-ic) [reply imas/ids_struct_array.py:132]: Hi Olivier,

> where the ids_path has slicing

This PR doesn't really solve that case either. I could still not do `wall.description_2d[0].vessel.unit[8:].element[:].name`: the slice at `unit[8:]` will return a list that has no attribute `element`.

> What would be the easiest/recommended way to go from an ids_path string that contains slicing indices to an actual list of objects?

This is actually a surprisingly tricky problem! And it mostly has to do with the possibility that IDSs don't have homogeneous array sizes deep down in the tree.

`ibex` has an implementation of this and, since it converts everything to JSON (which supports ragged arrays), it can avoid some of the difficulty. Looks like it is mostly implemented in [`_expand_single_path_element` in this file](https://git.iter.org/projects/IMEX/repos/ibex/browse/backend/ibex/data_source/imaspy_source.py?at=refs%2Fheads%2Fdevelop). Though quickly scanning it, it looks like it doesn't handle the negative indices either (which is probably fine for the `ibex` use case).

I see two options for implementing your use case, with option 1 the easiest to implement and option 2 more generally useful 🙂 
1. Add a method to [`IDSPath`](https://github.com/iterorganization/IMAS-Python/blob/develop/imas/ids_path.py), for example `IDSPath.goto_slices` that returns a list of elements that match the slice syntax. It would be an extended form of `IDSPath.goto` (which returns a single element and raises a `NotImplementedError` when a slice is used in the path),
2. Implement slices on arrays of structures, which return a new object (e.g. `IDSSlice`). The IDSSlice can keep track of all objects that matched the slice expression, and allow further slicing of child elements, getting the matching elements, etc. That could look as follows:
```python
>>> ids_slice = wall.description_2d[0].vessel.unit[8:]
>>> ids_slice
<IDSSlice (IDS:wall, description_2d[0].vessel.unit[8:], 4 matches)>
>>> ids_slice.element
<IDSSlice (IDS:wall, description_2d[0].vessel.unit[8:].element, 4 matches)>
>>> ids_slice.element[:]
<IDSSlice (IDS:wall, description_2d[0].vessel.unit[8:].element[:], 6 matches)>
>>> ids_slice.element[:].name
<IDSSlice (IDS:wall, description_2d[0].vessel.unit[8:].element[:].name, 6 matches)>
>>> list(ids_slice.element[:].name)
[<IDSString0D (IDS:wall, description_2d[0].vessel.unit[8].element[0].name)>,
 <IDSString0D (IDS:wall, description_2d[0].vessel.unit[8].element[1].name)>,
 <IDSString0D (IDS:wall, description_2d[0].vessel.unit[9].element[0].name)>,
 ...
 ]
```

In either case you'll need to decide how to handle edge cases. For example, what does it mean when you do `unit[:2].element[2]` when `unit[0]` has 4 elements and `unit[1]` has only 2 elements. Is this an IndexError? Or does this return a match of `[unit[0].element[2]]` and just ignore that there is no `unit[1].element[2]`?

There's probably more dragons hiding when implementing something like this -- there's a reason that it's not currently implemented in IDSPath 😉

Comment (olivhoenen) [reply imas/ids_struct_array.py:132]: I'm not sure why inhomogeneity is a problem there: if a user ask for a slice that does not exist we have an out-of-bound error, whether it is homogeneous or not (admittedly the risk increase in inhomogeneous case, but that should not be a concern for the implementation). IMO index-errors shall not be swallowed if they happen in nested elements, the error shall then just be propagated.

What do you think @prasad-sawantdesai, would something like 1 or 2 be useful in this case? 2 looks more appealing to me (but maybe that's because of the code snippet :roll_eyes: ).

Comment (prasad-sawantdesai) [reply imas/ids_struct_array.py:132]: I will try to implement second option :)

File: imas/ids_slice.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    - Attribute access on all matched elements
+    - Iteration over matched elements
+    """
+
+    __slots__ = ["_parent", "metadata", "_matched_elements", "_slice_path", "_lazy"]
+
+    def __init__(
+        self,
+        parent: IDSBase,
+        metadata: Any,
```
Comment (maarten-ic) [inline imas/ids_slice.py]:
```suggestion
        metadata: IDSMetadata,
```

File: imas/ids_slice.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+            if isinstance(element, IDSPrimitive):
+                # Extract the wrapped value from IDSPrimitive
+                result.append(element.value)
+            else:
+                # Return other types as-is (structures, arrays, etc.)
+                result.append(element)
+        return result
+
+    def flatten(self, recursive: bool = False) -> "IDSSlice":
+        """Flatten nested arrays into a single IDSSlice.
```
Comment (maarten-ic) [inline imas/ids_slice.py]: I don't understand how you can end up with nested arrays, can you clarify?

File: imas/ids_struct_array.py [hunk 4]
```diff
@@ -120,13 +120,65 @@ def _element_structure(self):
...
+            String representation like "[1:5]", "[::2]", etc.
+        """
+        start = slice_obj.start if slice_obj.start is not None else ""
+        stop = slice_obj.stop if slice_obj.stop is not None else ""
+        step = slice_obj.step if slice_obj.step is not None else ""
+
+        if step:
+            return f"[{start}:{stop}:{step}]"
+        else:
+            return f"[{start}:{stop}]"
```
Comment (maarten-ic) [inline imas/ids_struct_array.py]: DRY, make this the responsibility of the IDSSlice class only

File: imas/ids_struct_array.py [hunk 5]
```diff
@@ -120,13 +120,65 @@ def _element_structure(self):
...
+        """Get element(s) from the struct array.
+
+        Args:
+            item: Integer index or slice object
+
+        Returns:
+            A single IDSStructure if item is an int, or an IDSSlice if item is a slice
+        """
+        if isinstance(item, slice):
+            # Slicing is not supported on lazy-loaded arrays
```
Comment (maarten-ic) [inline imas/ids_struct_array.py]: Why not?

File: imas/ids_struct_array.py [hunk 6]
```diff
@@ -120,13 +120,65 @@ def _element_structure(self):
...
+                    "Use integer indexing instead."
+                )
+
+            # Handle slice by returning an IDSSlice
+            from imas.ids_slice import IDSSlice
+
+            # Get the matched elements
+            matched_elements = self.value[item]
+            if not isinstance(matched_elements, list):
+                matched_elements = [matched_elements]
```
Comment (maarten-ic) [inline imas/ids_struct_array.py]:
```suggestion
```

**Review by maarten-ic — COMMENTED** [review]
Hi Prasad, I had a quick look at the current state of the PR.

Please have a look at below comments, I think it can simplify this logic a lot 🙂

File: imas/ids_slice.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        # Avoid issues with special attributes
+        if name.startswith("_"):
+            raise AttributeError(f"IDSSlice has no attribute '{name}'")
+
+        # Access the attribute on each element
+        child_elements = []
+
+        for element in self._matched_elements:
+            child = getattr(element, name)
+            child_elements.append(child)
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Applied [156aded](https://github.com/iterorganization/IMAS-Python/pull/20/commits/156adedcb0f79d1ef9d746aad2f4d7f18bafc8bf)

File: imas/ids_slice.py (hunk 2 above)
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Updated [8398537](https://github.com/iterorganization/IMAS-Python/pull/20/commits/83985376349d4080ce42b03edfce146c87a95be7)

File: imas/ids_struct_array.py (hunk 6 above)
Comment (prasad-sawantdesai) [reply imas/ids_struct_array.py]: Applied [2423a89](https://github.com/iterorganization/IMAS-Python/pull/20/commits/2423a8901612547ec860c9a1e6fa0b84dfcb58ba)

**Review by maarten-ic — CHANGES_REQUESTED** [review]
Hi Prasad,

> 3. It retrieves the grid IDSStructure from all profiles_1d, and accessing the first element returns an IDSStructure
> [...]
> In the example below, cp.profiles_1d[1:3].ion[0] returns the first ion IDSStructure from all selected profiles_1d

Yes, I understand what it's doing, but I find it illogical and unexpected 😉 
I would prefer to follow [the principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), which - to me - would be: any operation on the IDSSlice should translate to the operation on the underlying IDS objects.

In pseudo code:
```python
def do_something_with(ids_slice):
    return IDSSlice([do_something_with(ids_object) for ids_object in ids_slice])
```

We're doing this in most cases now (e.g. `__getattr__`, `__getitem__` when the underlying objects are IDSStructArray) but not when the underlying objects are IDSStructure.

Let's set up a meeting to discuss in more detail if you want 🙂 

(I still haven't reviewed the code in more detail, sorry! I'll do that this week.)

File: imas/ids_slice.py [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+        except (ValueError, TypeError):
+            pass
+
+        # If reshape fails or not all elements are extractable, return as object array
+        try:
+            return np.array(flat_values, dtype=object).reshape(self._virtual_shape[0:1])
+        except (ValueError, TypeError):
+            return flat_values
+
+    def to_array(self) -> np.ndarray:
```
Comment (maarten-ic) [inline imas/ids_slice.py]: I would also simplify this method: tensorization is only relevant for when the slice refers to leaf nodes (e.g. FLT_1D).

Proposal:
- to_array raises an Exception when the IDSSlice refers to IDSStructure or IDSStructArrays
- to_array raises an Exception when the shape is irregular
- to_array never returns object arrays (when the user calls this method, they want a tensorized array, not object arrays)

**Review by maarten-ic — CHANGES_REQUESTED** [review]

File: imas/ids_struct_array.py [hunk 9]
```diff
@@ -121,12 +121,56 @@ def _element_structure(self):
...
+                from imas.ids_slice import IDSSlice
+
+                slice_str = IDSSlice._format_slice(item)
+
+                return IDSSlice(
+                    self.metadata,
+                    loaded_elements,
+                    slice_str,
+                    parent_array=self,
+                )
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_struct_array.py]: Thank you. updated f085f827cec632dd152d5dff3d50cef8f3f48bdd

File: imas/ids_slice.py (hunk 8 above)
Comment (prasad-sawantdesai) [reply imas/ids_slice.py:483]: Thank you.. updated as per suggestion and removed updating object arrays 44a111066e3af73d6157ddd266a340eed2f4fa24

File: imas/ids_slice.py [hunk 10]
```diff
@@ -0,0 +1,572 @@
...
+                    multi-dimensional slices. If False (default), return flat list
+                    or list of extracted values.
+
+        Returns:
+            list or numpy.ndarray: Extracted values as follows:
+
+            - 1D slices: List of raw Python/numpy values or unwrapped elements
+            - Multi-D with reshape=False: List of elements (each being an array)
+            - Multi-D with reshape=True: numpy.ndarray with shape self.shape,
+              or nested lists/object array representing structure
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Thank you for proposal.. Simplified values method 7d8abaae0229a5733afc9a55c08376b0c51f9e69

File: imas/ids_slice.py [hunk 11]
```diff
@@ -0,0 +1,572 @@
...
+        ids_name = "unknown"
+        full_path = self._path
+
+        if self._parent_array is not None:
+            ids_name = get_toplevel(self._parent_array).metadata.name
+            parent_array_path = get_full_path(self._parent_array)
+            full_path = parent_array_path + self._path
+        item_word = "item" if len(self) == 1 else "items"
+        my_repr += f" (IDS:{ids_name}, {full_path} with {len(self)} {item_word})>"
+        return my_repr
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Added ids_name property in IDSMetadata.. and passed full path when building slice.. 16a608a50cdb8492dbe4f452c69d5545625dc5d5

File: imas/ids_slice.py [hunk 12]
```diff
@@ -0,0 +1,572 @@
...
+        from imas.ids_primitive import IDSNumericArray
+
+        if isinstance(child_elements[0], IDSStructArray):
+            # Children are IDSStructArray - track the new dimension
+            child_sizes = [len(arr) for arr in child_elements]
+
+            # New virtual shape: current shape + new dimension
+            new_virtual_shape = self._virtual_shape + (
+                child_sizes[0] if child_sizes else 0,
+            )
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Thanks.. you are right.. Update the same 74e0d73071569b5fd59c9009aec38352aa2dec89

I agree that this PR is getting complicated. I will finish review comments and then will see how code look like and decide to create different PRs.

File: imas/ids_slice.py [hunk 13]
```diff
@@ -0,0 +1,572 @@
...
+        """
+        if not self._matched_elements:
+            raise IndexError(
+                f"Cannot access node '{name}' on empty slice with 0 elements"
+            )
+
+        from imas.ids_struct_array import IDSStructArray
+
+        child_metadata = None
+        if self.metadata is not None:
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Right, IDSSlice should always have valid metadata.. d91d25b4ce02a734bf6f7dc581c39ae77f772a0e

File: imas/ids_slice.py [hunk 14]
```diff
@@ -0,0 +1,572 @@
...
+        Args:
+            name: Name of the node to access
+
+        Returns:
+            A new IDSSlice containing the child node from each matched element
+        """
+        if not self._matched_elements:
+            raise IndexError(
+                f"Cannot access node '{name}' on empty slice with 0 elements"
+            )
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Thanks for checking this.. Fixed 0a4658d5b732b5c01f1002e3cfcafa1c0800a9f5

File: imas/ids_slice.py [hunk 15]
```diff
@@ -0,0 +1,572 @@
...
+
+    def __len__(self) -> int:
+        """Return the number of elements matched by this slice."""
+        return len(self._matched_elements)
+
+    def __iter__(self) -> Iterator[Any]:
+        """Iterate over all matched elements."""
+        return iter(self._matched_elements)
+
+    def __getitem__(self, item: Union[int, slice]) -> Union[Any, "IDSSlice"]:
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Simplified this method and it should return IDSSlice. b92883c238ac611741c37869bc5e8a3bd2dd5bd6

File: imas/ids_slice.py [hunk 16]
```diff
@@ -0,0 +1,572 @@
...
+    def shape(self) -> Tuple[int, ...]:
+        """Get the virtual multi-dimensional shape.
+
+        Returns the shape of the data as if it were organized in a multi-dimensional
+        array, based on the hierarchy of slicing operations performed.
+
+        Returns:
+            Tuple of dimensions.
+        """
+        return self._virtual_shape
```
[the remark this answers is not in the document]
Comment (prasad-sawantdesai) [reply imas/ids_slice.py]: Right.. modified this method 56f154ec03953d3139bfb463583b41097a4ba2bf

**Comment by prasad-sawantdesai** [comment]
> Hi Prasad,
> 
> > 3. It retrieves the grid IDSStructure from all profiles_1d, and accessing the first element returns an IDSStructure
> >    [...]
> >    In the example below, cp.profiles_1d[1:3].ion[0] returns the first ion IDSStructure from all selected profiles_1d
> 
> Yes, I understand what it's doing, but I find it illogical and unexpected 😉 I would prefer to follow [the principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), which - to me - would be: any operation on the IDSSlice should translate to the operation on the underlying IDS objects.
> 
> In pseudo code:
> 
> ```python
> def do_something_with(ids_slice):
>     return IDSSlice([do_something_with(ids_object) for ids_object in ids_slice])
> ```
> 
> We're doing this in most cases now (e.g. `__getattr__`, `__getitem__` when the underlying objects are IDSStructArray) but not when the underlying objects are IDSStructure.
> 
> Let's set up a meeting to discuss in more detail if you want 🙂
> 
> (I still haven't reviewed the code in more detail, sorry! I'll do that this week.)

Right.. Updated here a3ca48c8f94f06a5e3ada875e8b7853618b31d6c

File: imas/ids_slice.py [hunk 17]
```diff
@@ -0,0 +1,596 @@
+# This file is part of IMAS-Python.
+# You should have received the IMAS-Python LICENSE file with this project.
+"""IDSSlice represents a collection of IDS nodes matching a slice expression.
```
Comment (olivhoenen) [inline imas/ids_slice.py:3]: `IDSSlice` choice of name here is kind of misleading/confusing, as a "slice" of IDS is historically understood as a time slice.

File: imas/ids_slice.py [hunk 18]
```diff
@@ -0,0 +1,596 @@
...
+            full_path: Full path from the IDS root (e.g., "profiles_1d[:].ion[:]")
+            parent_array: Optional reference to the parent IDSStructArray for context
+            virtual_shape: Optional tuple representing multi-dimensional shape
+            element_hierarchy: Optional tracking of element grouping
+        """
+        self.metadata = metadata
+        self._matched_elements = matched_elements
+        self._slice_path = full_path
+        self._parent_array = parent_array
+        self._virtual_shape = virtual_shape or (len(matched_elements),)
```
Comment (olivhoenen) [inline imas/ids_slice.py:69]: What is the purpose of `_virtual_shape`? It does not seem to be used for anything.

File: imas/ids_slice.py [hunk 19]
```diff
@@ -0,0 +1,596 @@
...
+                full_path,
+                parent_array=self._parent_array,
+                virtual_shape=self._virtual_shape,
+                element_hierarchy=self._element_hierarchy,
+            )
+
+        # If matched_elements are IDSStructArray and we're accessing an
+        # attribute on them, we need to get that attribute from each
+        # array's elements
+        if isinstance(self._matched_elements[0], IDSStructArray):
```
Comment (olivhoenen) [inline imas/ids_slice.py:348]: Tested already in line 324, can probably be more logically structured

**Review by olivhoenen — COMMENTED** [review]
Sorry @prasad-sawantdesai my comments remained pending since months as I forgot to submit
